control-plane/agents/common/src/errors.rs

```diff
@@ -0,0 +1,300 @@
+                resource: ResourceKind::Pool,
+                source: desc.to_string(),
+                extra: error.full_string(),
+            },
+            SvcError::ReplicaNotFound {
+                ..
+            } => ReplyError {
+                kind: ReplyErrorKind::NotFound,
+                resource: ResourceKind::Replica,
+                source: desc.to_string(),
+                extra: error.full_string(),
+            },
+            SvcError::NexusNotFound {
+                ..
+            } => ReplyError {
+                kind: ReplyErrorKind::NotFound,
+                resource: ResourceKind::Nexus,
+                source: desc.to_string(),
+                extra: error.full_string(),
+            },
+            SvcError::InvalidFilter {
+                ..
+            } => ReplyError {
+                kind: ReplyErrorKind::Internal,
+                resource: ResourceKind::Unknown,
+                source: desc.to_string(),
+                extra: error.full_string(),
+            },
+            SvcError::Internal {
+                ..
+            } => ReplyError {
+                kind: ReplyErrorKind::Internal,
+                resource: ResourceKind::Unknown,
+                source: desc.to_string(),
+                extra: error.full_string(),
+            },
+            SvcError::MBusError {
+                source,
+            } => source.into(),
+        }
+    }
+}
+
+fn grpc_to_reply_error(error: SvcError) -> ReplyError {
+    match error {
+        SvcError::GrpcRequestError {
+            source,
+            request,
+            resource,
+        } => {
+            let kind = match source.code() {
+                Code::Ok => ReplyErrorKind::Internal,
+                Code::Cancelled => ReplyErrorKind::Internal,
+                Code::Unknown => ReplyErrorKind::Internal,
+                Code::InvalidArgument => ReplyErrorKind::InvalidArgument,
+                Code::DeadlineExceeded => ReplyErrorKind::DeadlineExceeded,
+                Code::NotFound => ReplyErrorKind::NotFound,
+                Code::AlreadyExists => ReplyErrorKind::AlreadyExists,
+                Code::PermissionDenied => ReplyErrorKind::PermissionDenied,
+                Code::ResourceExhausted => ReplyErrorKind::ResourceExhausted,
+                Code::FailedPrecondition => ReplyErrorKind::FailedPrecondition,
+                Code::Aborted => ReplyErrorKind::Aborted,
+                Code::OutOfRange => ReplyErrorKind::OutOfRange,
+                Code::Unimplemented => ReplyErrorKind::Unimplemented,
+                Code::Internal => ReplyErrorKind::Internal,
+                Code::Unavailable => ReplyErrorKind::Unavailable,
+                Code::DataLoss => ReplyErrorKind::Internal,
+                Code::Unauthenticated => ReplyErrorKind::Unauthenticated,
+                Code::__NonExhaustive => ReplyErrorKind::Internal,
+            };
+            let extra = format!("{}::{}", request, source.to_string());
```

Comment:
Is it worth implementing a From trait here to convert a Code to a ReplyErrorKind

Reply:
Yes I think so :+1:

Reply:
Oh right, we can't because both types aren't local to the crate...